### PR TITLE
Add ServiceChannel provider

### PR DESCRIPTION
**Pull request by dev-hyang**
1, Add ServiceChannel provider;
2, Add ServiceChannel Test case;

### CODE REVIEW

File: AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="RemoteAuthenticationOptions.CallbackPath"/>.
+        /// </summary>
+        public const string CallbackPath = "/signin-servicechannel";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.AuthorizationEndpoint"/>.
+        /// </summary>
+        public const string AuthorizationEndpoint = "https://login.servicechannel.com/oauth/authorize";
```
Comment (martincostello) [inline AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs]:
```suggestion
        public static readonly string AuthorizationEndpoint = "https://login.servicechannel.com/oauth/authorize";
```

File: AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        /// <summary>
+        /// Default value for <see cref="AuthenticationScheme.Name"/>.
+        /// </summary>
+        public const string AuthenticationScheme = "ServiceChannel";
+
+        /// <summary>
+        /// Default value for <see cref="AuthenticationScheme.DisplayName"/>.
+        /// </summary>
+        public const string DisplayName = "ServiceChannel";
```
Comment (martincostello) [inline AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs]:
```suggestion
        public static readonly string DisplayName = "ServiceChannel";
```

File: AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="AuthenticationSchemeOptions.ClaimsIssuer"/>.
+        /// </summary>
+        public const string Issuer = "ServiceChannel";
```
Comment (martincostello) [inline AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs]:
```suggestion
        public static readonly string Issuer = "ServiceChannel";
```

File: AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="RemoteAuthenticationOptions.CallbackPath"/>.
+        /// </summary>
+        public const string CallbackPath = "/signin-servicechannel";
```
Comment (martincostello) [inline AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs]:
```suggestion
        public static readonly string CallbackPath = "/signin-servicechannel";
```

File: AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.AuthorizationEndpoint"/>.
+        /// </summary>
+        public const string AuthorizationEndpoint = "https://login.servicechannel.com/oauth/authorize";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
+        /// </summary>
+        public const string TokenEndpoint = "https://login.servicechannel.com/oauth/token";
```
Comment (martincostello) [inline AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs]:
```suggestion
        public static readonly string TokenEndpoint = "https://login.servicechannel.com/oauth/token";
```

File: AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
+        /// </summary>
+        public const string UserInformationEndpoint = "https://api.servicechannel.com/v3/users/current/profile";
```
Comment (martincostello) [inline AspNet.Security.OAuth.ServiceChannel/ServiceChannelAuthenticationDefaults.cs]:
```suggestion
        public static readonly string UserInformationEndpoint = "https://api.servicechannel.com/v3/users/current/profile";
```

File: samples/Mvc.Client/Mvc.Client.csproj [hunk 7]
```diff
@@ -14,6 +14,10 @@
     <PackageReference Include="Microsoft.AspNetCore.Authentication.Twitter" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\AspNet.Security.OAuth.ServiceChannel\AspNet.Security.OAuth.ServiceChannel.csproj" />
+  </ItemGroup>
+
```
Comment (martincostello) [inline samples/Mvc.Client/Mvc.Client.csproj]: This isn't needed as it's added automatically by lines 8-10.
```suggestion
```

File: test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj [hunk 8]
```diff
@@ -39,4 +39,8 @@
     </AssemblyAttribute>
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\AspNet.Security.OAuth.ServiceChannel\AspNet.Security.OAuth.ServiceChannel.csproj" />
+  </ItemGroup>
+
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj]: Already handled by lines 17-19.

```suggestion
```

**Review by martincostello — COMMENTED** [review]
Please add the provider to the table in the README file.

**Comment by dev-hyang** [comment]
Suppose the wildcard of project group in Tests and Mvc.Client project would work? But it seems that the reference is not valid after removed.

**Comment by martincostello** [comment]
Ah - it's because you haven't put the project files for the provider in the right place - they should be in the `src/AspNet.Security.OAuth.ServiceChannel` directory, not in the root directory of the solution.

File: samples/Mvc.Client/Startup.cs [hunk 9]
```diff
@@ -61,6 +61,13 @@ public void ConfigureServices(IServiceCollection services)
                 options.Scope.Add("user:email");
             })
 
+            .AddServiceChannel(options =>
+            {
+                options.UserInformationEndpoint = Configuration["ServiceChannel:UserInformationEndpoint"];
+                options.ClientId = Configuration["ServiceChannel:ClientId"];
+                options.ClientSecret = Configuration["ServiceChannel:ClientSecret"];
+            })
+
```
Comment (martincostello) [inline samples/Mvc.Client/Startup.cs]: Adding every new provider to our samples is not sustainable.

```suggestion
```

File: samples/Mvc.Client/Properties/launchSettings.json [hunk 10]
```diff
@@ -18,10 +18,11 @@
     "web": {
       "commandName": "Project",
       "launchBrowser": true,
-      "launchUrl": "https://localhost:5001/",
+      "launchUrl": "https://localhost:6001/",
```
Comment (martincostello) [inline samples/Mvc.Client/Properties/launchSettings.json]:
```suggestion
      "launchUrl": "https://localhost:5001/",
```

File: samples/Mvc.Client/Properties/launchSettings.json [hunk 11]
```diff
@@ -18,10 +18,11 @@
...
       "environmentVariables": {
         "ASPNETCORE_ENVIRONMENT": "Development"
-      }
+      },
+      "applicationUrl": "https://localhost:6001/"
```
Comment (martincostello) [inline samples/Mvc.Client/Properties/launchSettings.json]:
```suggestion
      "applicationUrl": "https://localhost:5001/"
```

File: samples/Mvc.Client/Properties/launchSettings.json [hunk 12]
```diff
@@ -18,10 +18,11 @@
...
     }
   }
-}
+}
```
Comment (martincostello) [inline samples/Mvc.Client/Properties/launchSettings.json:28]:
```suggestion
}

```

File: samples/Mvc.Client/appsettings.json [hunk 13]
```diff
@@ -14,5 +14,10 @@
   "Twitter": {
     "ConsumerKey": "6XaCTaLbMqfj6ww3zvZ5g",
     "ConsumerSecret": "Il2eFzGIrYhz6BWjYhVXBPQSfZuS4xoHpSSyD9PI"
+  },
+  "ServiceChannel": {
+    "UserInformationEndPoint": "https://sb2api.servicechannel.com/v3/users/current/profile",
+    "ClientId": "PR.2000156703.19A2F24F-ABB2-437A-B527-FC18493D9A4E",
+    "ClientSecret": "F0635EA2-2417-440B-88D9-7E2BD6208287"
```
Comment (martincostello) [inline samples/Mvc.Client/appsettings.json]: Please remove, and if these are real credentials you should revoke/rotate them.

```suggestion
```

**Comment by dev-hyang** [comment]
Solved in the commit "Move directory and recover Mvc.Client launch setting"

**Comment by martincostello** [comment]
I assume you have because of the changes to the client app, before before I merge this can you just confirm that you've tested these changes with the real ServiceChannel service and that the provider works as expected with it?

**Comment by dev-hyang** [comment]
> I assume you have because of the changes to the client app, before before I merge this can you just confirm that you've tested these changes with the real ServiceChannel service and that the provider works as expected with it?

Yes, I did test it with the ServiceChannel Sandbox Environment and it works well.

**Comment by martincostello** [comment]
Thanks again for your contribution - this provider is now available from NuGet.org: https://www.nuget.org/packages/AspNet.Security.OAuth.ServiceChannel/5.0.17

**Pull request merged**